### PR TITLE
Bg_Iknv_Doukutu

### PR DESCRIPTION
**Pull request by Kelebek1**

### CODE REVIEW

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 1]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD72BC.s")
+    if ((BGIKNVDOUKUTU_GET_F(&this->dyna.actor) == 1) || (BGIKNVDOUKUTU_GET_F(&this->dyna.actor) == 2)) {
+        DynaPoly_DeleteBgActor(play, &play->colCtx.dyna, this->dyna.bgId);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD7360.s")
+void func_80BD716C(BgIknvDoukutu* this, PlayState* play) {
+    if (this->unk_160 >= 0.05f) {
+        this->unk_160 = this->unk_160 - 0.05f;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:108]:
```suggestion
        this->unk_160 -= 0.05f;
```

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 2]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD7538.s")
+void func_80BD72BC(BgIknvDoukutu* this, PlayState* play) {
+    if (Cutscene_CheckActorAction(play, this->unk_15E) &&
+        (play->csCtx.actorActions[Cutscene_GetActorActionIndex(play, this->unk_15E)]->action == 3)) {
+        this->actionFunc = func_80BD7250;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD7768.s")
+    if (this->unk_160 < 0.95f) {
+        this->unk_160 = this->unk_160 + 0.05f;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:143]:
```suggestion
        this->unk_160 += 0.05f;
```

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 3]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
+
+    func_80BD7768(&this->dyna.actor, play);
+
+    gfxCtx = play->state.gfxCtx;
+    OPEN_DISPS(gfxCtx);
+    f32 sp48;
+    f32 sp44;
+    f32 sp40;
+    s16 sp3E = Math_Vec3f_Yaw(&play->cameraPtrs[play->activeCamera]->eye, &play->cameraPtrs[play->activeCamera]->at);
+    s16 sp3C = -Math_Vec3f_Pitch(&play->cameraPtrs[play->activeCamera]->eye, &play->cameraPtrs[play->activeCamera]->at);
```
Comment (hensldm) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:198]:
```suggestion
    s16 sp3E = Math_Vec3f_Yaw(&GET_ACTIVE_CAM(play)->eye, &GET_ACTIVE_CAM(play)->at);
    s16 sp3C = -Math_Vec3f_Pitch(&GET_ACTIVE_CAM(play)->eye, &GET_ACTIVE_CAM(play)->at);
```

**Review by hensldm — COMMENTED** [review]
A few minor things.

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 4]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
     (ActorFunc)BgIknvDoukutu_Draw,
 };
 
-#endif
+void BgIknvDoukutu_Init(Actor* thisx, PlayState* play) {
+    BgIknvDoukutu* this = THIS;
+    CollisionHeader* sp30 = NULL;
+    s32 pad;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:45]:
```suggestion
    s32 pad;
    BgIknvDoukutu* this = THIS;
    CollisionHeader* sp30 = NULL;
```

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 5]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
-extern UNK_TYPE D_060117A0;
-extern UNK_TYPE D_06012728;
-extern UNK_TYPE D_06012788;
+    Actor_SetScale(&this->dyna.actor, 0.1f);
+    this->actionFunc = func_80BD73D0;
+    this->unk_15C = 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/BgIknvDoukutu_Init.s")
+    switch (BGIKNVDOUKUTU_GET_F(&this->dyna.actor)) {
+        case 0:
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:52]: Enum for these cases.

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 6]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
     (ActorFunc)BgIknvDoukutu_Draw,
 };
 
-#endif
+void BgIknvDoukutu_Init(Actor* thisx, PlayState* play) {
+    BgIknvDoukutu* this = THIS;
+    CollisionHeader* sp30 = NULL;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:44]: `colHeader`

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 7]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD7768.s")
+    if (this->unk_160 < 0.95f) {
+        this->unk_160 = this->unk_160 + 0.05f;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Iknv_Doukutu/func_80BD7820.s")
+void func_80BD7360(BgIknvDoukutu* this, PlayState* play) {
+    if (Cutscene_CheckActorAction(play, this->unk_15E) &&
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:148]: `csAction`

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 8]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
+    GraphicsContext* gfxCtx;
+    f32 sp54;
+
+    func_80BD7768(&this->dyna.actor, play);
+
+    gfxCtx = play->state.gfxCtx;
+    OPEN_DISPS(gfxCtx);
+    f32 sp48;
+    f32 sp44;
+    f32 sp40;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:196]: Vec3f?

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c [hunk 9]
```diff
@@ -35,38 +39,234 @@ const ActorInit Bg_Iknv_Doukutu_InitVars = {
...
+
+    gSPDisplayList(POLY_XLU_DISP++, object_iknv_obj_DL_012700);
+
+    CLOSE_DISPS(play->state.gfxCtx);
+}
+
+void func_80BD78C4(Actor* thisx, PlayState* play) {
+    BgIknvDoukutu* this = THIS;
+    f32 sp30 = this->unk_160;
+    s32 pad;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:257]:
```suggestion
    s32 pad;
    BgIknvDoukutu* this = THIS;
    f32 sp30 = this->unk_160;
```

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.h [hunk 10]
```diff
@@ -7,11 +7,15 @@ struct BgIknvDoukutu;
...
 
+#define BGIKNVDOUKUTU_GET_F(thisx) ((thisx)->params & 0xF)
+
 typedef struct BgIknvDoukutu {
-    /* 0x0000 */ Actor actor;
-    /* 0x0144 */ char unk_144[0x20];
-    /* 0x0164 */ BgIknvDoukutuActionFunc actionFunc;
-} BgIknvDoukutu; // size = 0x168
+    /* 0x000 */ DynaPolyActor dyna;
+    /* 0x15C */ s16 unk_15C;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.h:14]:
```suggestion
    /* 0x15C */ s16 unk_15C; // set and not used
```

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c (hunk 4 above)
Comment (Kelebek1) [reply src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:45]: Breaks stack.

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c (hunk 8 above)
Comment (Kelebek1) [reply src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:196]: Maybe, I wasn't really sure.

File: src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c (hunk 9 above)
Comment (Kelebek1) [reply src/overlays/actors/ovl_Bg_Iknv_Doukutu/z_bg_iknv_doukutu.c:257]: Breaks stack.

**Review by hensldm — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**